File: CCC Solutions/2000/CCC 00 S5 - Sheep and Coyotes (ccc00s5).py
n = int(input())
s = []
for _ in range(n):
    x, y = [float(input()) for _ in range(2)]
    s.append([x, y, [0.0, 1000.0]])

for i in range(n):
    xi, yi = s[i][0], s[i][1]
    li, ri = s[i][2]
    for j in range(n):
        if i == j:
            continue
        xj, yj = s[j][0], s[j][1]
        A = 2 * (xj - xi)
        B = xi**2 - xj**2 + yi**2 - yj**2
        if abs(A) < 1e-8:
            if B > 1e-8:
                li, ri = 1.0, 0.0
                break
            continue
        x0 = -B / A
        if A > 0:
            ri = min(ri, x0)
        else:
            li = max(li, x0)
        if li - ri > 1e-8:
            break
    s[i][2] = [li, ri]

for si in s:
    li, ri = si[2]
    li = max(li, 0.0)
    ri = min(ri, 1000.0)
    if ri - li >= -1e-8:
        print(f"The sheep at ({int(si[0])}, {int(si[1])}) might be eaten.")

File: CCC Solutions/2000/CCC 00 S5 - Sheep and Coyotes (ccc00s5).cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    int n;
    cin >> n;
    pair<long double,long double> s[1000];
    for (int i=0;i<n;i++){
        long double x,y;
        cin >> x >> y;
        s[i] = {x,y};
    }

    for (int i=0;i<n;i++){
        long double l=0,r=1000;
        for (int j=0;j<n;j++){
            if (i==j) continue;
            if (s[i].first==s[j].first){
                if (s[i].second<s[j].second) continue;
                else goto next;
            }
            long double midX=(s[i].first+s[j].first)/2, midY=(s[i].second+s[j].second)/2;
            long double split;
            if (s[i].second==s[j].second) split=midX;
            else split=-(midY - (-(s[i].first-s[j].first)/(s[i].second-s[j].second))*midX) / (-(s[i].first-s[j].first)/(s[i].second-s[j].second));
            if (s[i].first<s[j].first && r>split) r=split;
            if (s[i].first>s[j].first && split>l) l=split;
        }
        if (l<=r) printf("The sheep at (%.2Lf, %.2Lf) might be eaten.\n", s[i].first, s[i].second);
        next:;
    }
}


